Spring Data REST (#13373)
quarkusio/quarkus: docs/src/main/asciidoc/spring-data-rest.adoc

Quarkus provides a compatibility layer for Spring Data REST in a form of `spring-data-rest` extension.

include::./status-include.adoc[]

== Prerequisites

To complete this guide, you need:

* less than 15 minutes
* an IDE
* JDK 1.8+ installed with `JAVA_HOME` configured appropriately
* Apache Maven {maven-version}

== Solution

We recommend that you follow the instructions in the next sections and create the application step by step.
However, you can go right to the completed example.

Clone the Git repository: `git clone {quickstarts-clone-url}`, or download an {quickstarts-archive-url}[archive].

The solution is located in the `spring-data-rest-quickstart` {quickstarts-tree-url}/spring-data-rest-quickstart[directory].

== Creating the Maven project

First, we need a new project. Create a new project with the following command:

[source,bash,subs=attributes+]
----
mvn io.quarkus:quarkus-maven-plugin:{quarkus-version}:create \
    -DprojectGroupId=org.acme \
    -DprojectArtifactId=spring-data-rest-quickstart \
    -Dextensions="spring-data-rest,resteasy-jackson,quarkus-jdbc-postgresql"
cd spring-data-rest-quickstart
----

> This command generates a Maven project with a `spring-data-rest` extension.

Reviewer: I think you should mention the other extensions here too. Especially since you mention Jackson/JSON-B later.

----
package org.acme.spring.data.rest;

import javax.persistence.Entity;
import javax.persistence.GeneratedValue;
import javax.persistence.Id;

@Entity
public class Fruit {

    @Id
    @GeneratedValue
    private Long id;

    private String name;

    private String color;

    public Fruit() {
    }

    public Fruit(String name, String color) {
        this.name = name;
        this.color = color;
    }

    public Long getId() {
        return id;
    }

    public void setId(Long id) {
        this.id = id;
    }

    public String getName() {
        return name;
    }

    public void setName(String name) {
        this.name = name;
    }

    public String getColor() {
        return color;
    }

    public void setColor(String color) {
        this.color = color;
    }
}
----

== Configure database access properties

Add the following properties to `application.properties` to configure access to a local PostgreSQL instance.

[source,properties]
----
quarkus.datasource.db-kind=postgresql
quarkus.datasource.username=quarkus_test
quarkus.datasource.password=quarkus_test
quarkus.datasource.jdbc.url=jdbc:postgresql:quarkus_test
quarkus.datasource.jdbc.max-size=8

> quarkus.datasource.jdbc.min-size=2

Reviewer: Just remove this, it will generate a warning if you don't have the `initial-size` set.

A very easy way to accomplish that is by using the following Docker command:

[source,bash]
----
docker run --ulimit memlock=-1:-1 -it --rm=true --memory-swappiness=0 --name quarkus_test -e POSTGRES_USER=quarkus_test -e POSTGRES_PASSWORD=quarkus_test -e POSTGRES_DB=quarkus_test -p 5432:5432 postgres:11.5
----

If you plan on using a different setup, please change your `application.properties` accordingly.

== Prepare the data

To make it easier to showcase some capabilities of Spring Data REST on Quarkus, some test data should be inserted into the database
by adding the following content to a new file named `src/main/resources/import.sql`:

[source,sql]
----
INSERT INTO fruit(id, name, color) VALUES (1, 'Cherry', 'Red');
INSERT INTO fruit(id, name, color) VALUES (2, 'Apple', 'Red');
INSERT INTO fruit(id, name, color) VALUES (3, 'Banana', 'Yellow');
INSERT INTO fruit(id, name, color) VALUES (4, 'Avocado', 'Green');
INSERT INTO fruit(id, name, color) VALUES (5, 'Strawberry', 'Red');
----

Hibernate ORM will execute these queries on application startup.

== Define the repository

It is now time to define the repository that will be used to access `Fruit`.

> In a typical Spring Data fashion create a repository like so:

Reviewer: ```suggestion
In a typical Spring Data fashion, create a repository like so:
```

[source,java]
----
package org.acme.spring.data.rest;

import org.springframework.data.repository.CrudRepository;

public interface FruitsRepository extends CrudRepository<Fruit, Long> {
}
----

The `FruitsRepository` above extends Spring Data's `org.springframework.data.repository.CrudRepository` which means that all of the latter's methods are
available to `FruitsRepository`.

> `spring-data-jpa` will generate an implementation for this repository. Then `spring-data-rest` will generate a REST CRUD resource for it.

Reviewer 2: Is there a way to opt out of REST Resource generation?

Should there be some kind of build time warning if there is a clashing user defined Resource?

Author: You can annotate interface or a method with either `@RepositoryRestResource(exported=false)` or `@RestResource(exported=false)`. It will still generate a JAX-RS class but it will not have any method. I don't think that should cause any clashing though.

I'm not sure if there is an easy way to produce such a warning in a current state (there might be with the new REST extension). It is a valid scenario to have two classes on a same base path. In any case that would be something to look at in the `rest-data-panache` extension. `spring-data-rest` does not generate JAX-RS endpoints.

Reviewer 2: Good to know.

Does spring-data-rest proper have a way of disabling the exposure of the REST endpoint or is it all-or-nothing on the sense that if you have spring-data-rest, you expose all the repositories?

Author: By default Spring Data REST picks up almost everything everything. Then you can either disable separate repositories using the annotations that I've mentioned above or change the detection strategy in application properties (all, public, public or annotated (default), annotated).
I think we should add a similar configuration option too. But we could do it as a separate feature and add it to the `rest-data-panache` too. For now the only way to disable would be with an annotation.

Reviewer 2: Okay. Let's address the configuration in a follow-up PR

> `spring-data-jpa` will generate an implementation for this repository. Then `spring-data-rest` will generate a REST CRUD resource for it.

Reviewer: ```suggestion
The `spring-data-jpa` extension will generate an implementation for this repository. Then the `spring-data-rest` extension will generate a REST CRUD resource for it.
```

                .when().get("/fruits")
                .then()
                .statusCode(200)
                .body(
                        not(containsString("Cherry")),
                        containsString("Apple"),
                        containsString("Banana")
                );

        //Create a new Fruit
        given()
                .contentType("application/json")
                .accept("application/json")
                .body("{\"name\": \"Orange\", \"color\": \"Orange\"}")
                .when().post("/fruits")
                .then()
                .statusCode(201)
                .body(containsString("Orange"))
                .body("id", notNullValue())
                .extract().body().jsonPath().getString("id");

        //List all, Orange should be present now:
        given()
                .accept("application/json")
                .when().get("/fruits")
                .then()
                .statusCode(200)
                .body(
                        not(containsString("Cherry")),
                        containsString("Apple"),
                        containsString("Orange")
                );
    }
}

----

> The test can be easily run by issuing: `./mvnw test`

Reviewer: ```suggestion
The test can be easily run by issuing: `./mvnw test`.
```

== Package and run the application

Quarkus dev mode works with the defined repositories just like with any other Quarkus extension, greatly enhancing your productivity during the dev cycle.
The application can be started in dev mode as usual using:

[source, bash]
----

> ./mvnw compile quarkus:dev

Reviewer: ```suggestion
./mvnw quarkus:dev
```

----

== Run the application as a native binary

> You can of course create a native executable following the instructions of the link:building-native-image[this] guide.

Reviewer: ```suggestion
You can of course create a native executable following the instructions of the link:building-native-image[Building native executables] guide.
```

== Supported Spring Data REST functionalities

Quarkus currently supports a subset of Spring Data REST features, namely the most useful and most commonly used features.

=== What is supported

> The following sections described the most important supported features of Spring Data REST.

Reviewer: ```suggestion
The following sections describe the most important supported features of Spring Data REST.
```

==== Automatic REST endpoint generation

Interfaces that extend any of the following Spring Data repositories get automatically generated REST endpoints:

* `org.springframework.data.repository.CrudRepository`
* `org.springframework.data.repository.PagingAndSortingRepository`
* `org.springframework.data.jpa.repository.JpaRepository`

Endpoints generated from the above repositories expose five common REST operations:

* `GET /fruits` - lists all entities or returns a page if `PagingAndSortingRepository` or `JpaRepository` is used.
* `GET /fruits/:id` - returns an entity by ID.
* `POST /fruits` - creates a new entity.
* `PUT /fruits/:id` - updates an existing entity or creates a new one with a specified ID (if allowed by the entity definition).
* `DELETE /fruits/:id` - deletes an entity by ID.

There are two supported data types: `application/json` and `application/hal+json`.

> The former is used by default, but it is highly recommended to specify which one do you prefer with an `Accept` header.

Reviewer: ```suggestion
The former is used by default, but it is highly recommended to specify which one you prefer with an `Accept` header.
```

==== Exposing many entities

If a database contains many entities, it might not be a great idea to return them all at once.

> `PagingAndSortingRepository` allows `spring-data-rest` extension to access data in chunks.

Reviewer: ```suggestion
`PagingAndSortingRepository` allows the `spring-data-rest` extension to access data in chunks.
```

[source,java]
----
package org.acme.spring.data.rest;

import org.springframework.data.repository.PagingAndSortingRepository;

public interface FruitsRepository extends PagingAndSortingRepository<Fruit, Long> {
}
----

Now the `GET /fruits` will accept three new query parameters: `sort`, `page` and `size`.

|===
| Query parameter | Description | Default value | Example values

| `sort`
| Sorts the entities that are returned by the list operation
| ""
| `?sort=name` (ascending name), `?sort=name,-color` (ascending name and descending color)

| `page`
| Zero indexed page number. Invalid value is interpreted as 0.
| 0
| 0, 11, 100

| `size`
| Page size. Minimal accepted value is 1. Any lower value is interpreted as 1.
| 20
| 1, 11, 100
|===

For paged responses, `spring-data-rest` also returns a set of link headers that can be used to access other pages: first, previous, next and last.

==== Fine tuning endpoints generation

This allows user to specify which methods should be exposed and what path should be used to access them.
Spring Data REST provides two annotations that can be used: `@RepositoryRestResource` and `@RestResource`.

> `spring-data-rest` extension supports the `exported` and `path` attributes of these annotations.

Reviewer: ```suggestion
The `spring-data-rest` extension supports the `exported` and `path` attributes of these annotations.
```

> Assume for example that fruits repository should be accessible by a `/my-fruits` path and only allow `GET` operation.

Reviewer: ```suggestion
Assume for example that the Fruits repository should be accessible by a `/my-fruits` path and only allow `GET` operations.
```

import java.util.Optional;

import org.springframework.data.repository.CrudRepository;
import org.springframework.data.rest.core.annotation.RepositoryRestResource;
import org.springframework.data.rest.core.annotation.RestResource;

@RepositoryRestResource(exported = false, path = "/my-fruits")
public interface FruitsRepository extends CrudRepository<Fruit, Long> {

    @RestResource(exported = true)
    Optional<Fruit> findById(Long id);

    @RestResource(exported = true)
    Iterable<Fruit> findAll();
}
----

`spring-data-rest` uses only a subset of the repository methods for data access.
It is important to annotate the correct method in order to customize its REST endpoint:

|===
|REST operation |CrudRepository |PagingAndSortingRepository and JpaRepository

|Get by ID
|`Optional<T> findById(ID id)`
|`Optional<T> findById(ID id)`

|List
|`Iterable<T> findAll()`
|`Page<T> findAll(Pageable pageable)`

|Create
|`<S extends T> S save(S entity)`
|`<S extends T> S save(S entity)`

|Update
|`<S extends T> S save(S entity)`
|`<S extends T> S save(S entity)`

|Delete
|`void deleteById(ID id)`
|`void deleteById(ID id)`
|===

=== What is currently unsupported

> * Only the repository methods listed above are supported. No other standard nor custom methods are supported.

Reviewer 2: s/nor/or

* Only the `exposed` and `path` REST resource annotation properties are supported.

== Important Technical Note

Please note that the Spring support in Quarkus does not start a Spring Application Context nor are any Spring infrastructure classes run.
Spring classes and annotations are only used for reading metadata and / or are used as user code method return types or parameter types.

== More Spring guides

Quarkus has more Spring compatibility features. See the following guides for more details:

* link:spring-data-jpa[Quarkus - Extension for Spring Data JPA]
* link:spring-di[Quarkus - Extension for Spring DI]
* link:spring-web[Quarkus - Extension for Spring Web]
* link:spring-security[Quarkus - Extension for Spring Security]
* link:spring-cloud-config-client[Quarkus - Reading properties from Spring Cloud Config Server]
* link:spring-boot-properties[Quarkus - Extension for Spring Boot properties]
* link:spring-cache[Quarkus - Extension for Spring Cache]

> * link:spring-scheduled[Quarkus - Extension for Spring Scheduled]

Reviewer: This guide should be added as a link in all the other guides.
